chain/epoch_manager/src/lib.rs

```diff
@@ -288,59 +292,133 @@ impl EpochManager {
             }
             let block_stats = block_validator_tracker
                 .get(&(i as u64))
-                .unwrap_or_else(|| &ValidatorStats { expected: 0, produced: 0 });
-            // Note, validator_kickout_threshold is 0..100, so we use * 100 to keep this in integer space.
-            if block_stats.produced * 100
-                < u64::from(block_producer_kickout_threshold) * block_stats.expected
-            {
-                validator_kickout.insert(
-                    account_id.clone(),
-                    ValidatorKickoutReason::NotEnoughBlocks {
-                        produced: block_stats.produced,
-                        expected: block_stats.expected,
-                    },
-                );
-            }
+                .unwrap_or_else(|| &ValidatorStats { expected: 0, produced: 0 })
+                .clone();
             let mut chunk_stats = ValidatorStats { produced: 0, expected: 0 };
             for (_, tracker) in chunk_validator_tracker.iter() {
                 if let Some(stat) = tracker.get(&(i as u64)) {
                     chunk_stats.expected += stat.expected;
                     chunk_stats.produced += stat.produced;
                 }
             }
-            if chunk_stats.produced * 100
-                < u64::from(chunk_producer_kickout_threshold) * chunk_stats.expected
+            total_stake += v.stake();
+            validator_block_chunk_stats
+                .insert(account_id.clone(), BlockChunkValidatorStats { block_stats, chunk_stats });
+        }
+
+        // We want to make sure the total stake of validators that will be kicked out in this epoch doesn't exceed
+        // config.validator_max_kickout_stake_ratio of total stake.
+        // To achieve that, we sort all validators by their average uptime (average of block and chunk
+        // uptime) and add validators to `exempted_validators` one by one, from high uptime to low uptime,
+        // until the total excepted stake exceeds the ratio of total stake that we need to keep.
+        // Later when we perform the check to kick out validators, we don't kick out validators in
+        // exempted_validators.
+        #[allow(unused_mut)]
+        let mut exempted_validators = HashSet::<&AccountId>::new();
+        #[cfg(feature = "protocol_feature_max_kickout_stake")]
+        if checked_feature!(
```

Comment:
nit: consider moving the computation of the exempted_validators to a separate method